### PR TITLE
Correct response codes and messages related to private factory flow

### PR DESCRIPTION
**Pull request by mshaposhnik**
### What does this PR do?
To support OAuth flow for private repository factories,  Che server must return correct responses to the clients. Most 
important use case is initiating the OAuth procedure. 

 - User accepts factory from the private repository;

 - Dashboard making request to resolve factory details using the repository URL which is provided by user;
(POST /api/factory/resolver)

- Che server recognizes that URL points to the private repository and looks for the credentials secret; If no secret present,
initiates the OAuth login procedure by returning the following response:

`POST https://<che-host>/api/factory/resolver`:

```
Response Status: 401

Response body:

 {
 "attributes": {
 "oauth_provider": "<provider>",
 "oauth_version": "<1.0|2.0>"
 "oauth_authentication_url": "https://<che-host>/api/oauth/<1.0>/authenticate?oauth_provider=<provider>&request_method=<GET|POST>&signature_method=<rsa|ssha>"
 },
 "errorCode": 401,
 "message": "SCM Authentication required"
}
```

Please note that parameters `token` and `redirect_after_login` are not returned and must be set by the client at its discretion.

Other use cases responses:

OAuth not configured:

```
Response Status:  500

Message: Provided location is unknown or misconfigured on the server-side. Error message: <>
```

Cannot communicate with OAuth server:

```
Response Status:  500

Message: There is an error that happened when communicating with SCM server. Error message: <>
```

### What issues does this PR fix or reference?

https://github.com/eclipse/che/issues/18389

### How to test this PR?

Run Che with bitbucket OAuth enabled and simulate different situations (user unauthorized etc)

### PR Checklist

[As the author of this Pull Request I made sure that:](https://github.com/eclipse/che/blob/master/CONTRIBUTING.md#pull-request-template-and-its-checklist)

- [x] [The Eclipse Contributor Agreement is valid](https://github.com/eclipse/che/blob/master/CONTRIBUTING.md#the-eclipse-contributor-agreement-is-valid)
- [x] [Code produced is complete](https://github.com/eclipse/che/blob/master/CONTRIBUTING.md#code-produced-is-complete)
- [x] [Code builds without errors](https://github.com/eclipse/che/blob/master/CONTRIBUTING.md#code-builds-without-errors)
- [x] [Tests are covering the bugfix](https://github.com/eclipse/che/blob/master/CONTRIBUTING.md#tests-are-covering-the-bugfix)
- [x] [The repository devfile is up to date and works](https://github.com/eclipse/che/blob/master/CONTRIBUTING.md#the-repository-devfile-is-up-to-date-and-works)
- [x] [Sections `What issues does this PR fix or reference` and `How to test this PR` completed](https://github.com/eclipse/che/blob/master/CONTRIBUTING.md#sections-what-issues-does-this-pr-fix-or-reference-and-how-to-test-this-pr-completed)
- [ ] [Relevant user documentation updated](https://github.com/eclipse/che/blob/master/CONTRIBUTING.md#relevant-contributing-documentation-updated)
- [ ] [Relevant contributing documentation updated](https://github.com/eclipse/che/blob/master/CONTRIBUTING.md#relevant-contributing-documentation-updated)
- [ ] [CI/CD changes implemented, documented and communicated](https://github.com/eclipse/che/blob/master/CONTRIBUTING.md#cicd-changes-implemented-documented-and-communicated)

### Reviewers

Reviewers, please comment how you tested the PR when approving it.

### CODE REVIEW

File: wsmaster/che-core-api-auth/src/main/java/org/eclipse/che/security/oauth1/OAuthAuthenticator.java [hunk 1]
```diff
@@ -229,6 +229,14 @@ String callback(final URL requestUrl) throws OAuthAuthenticationException {
    */
   abstract String getOAuthProvider();
 
+  /**
+   * Returns URL to initiate authentication process using given authenticator. Typically points to
+   * {@code /api/oauth/} or {@code /api/oauth/1.0} endpoint with necessary request params.
+   *
+   * @return URL to initiate authentication process
+   */
+  public abstract String getLocalAuthenticateUrl();
```
Comment (skabashnyuk) [inline wsmaster/che-core-api-auth/src/main/java/org/eclipse/che/security/oauth1/OAuthAuthenticator.java:238]: what was the motivation to put `local` in the name? What about `getAuthenticateUrl`?

**Review by skabashnyuk — COMMENTED** [review]
In general looks good

File: wsmaster/che-core-api-auth/src/main/java/org/eclipse/che/security/oauth1/OAuthAuthenticator.java (hunk 1 above)
Comment (mshaposhnik) [reply wsmaster/che-core-api-auth/src/main/java/org/eclipse/che/security/oauth1/OAuthAuthenticator.java:238]: there is already one for remote URL-s
https://github.com/eclipse/che/blob/108be5cd84b8535e38f3ac500e36e1fbc2e4f4e6/wsmaster/che-core-api-auth/src/main/java/org/eclipse/che/security/oauth1/OAuthAuthenticator.java#L102-L106

Comment (skabashnyuk) [reply wsmaster/che-core-api-auth/src/main/java/org/eclipse/che/security/oauth1/OAuthAuthenticator.java:238]: ok. and? 
What exactly means local?

Comment (mshaposhnik) [reply wsmaster/che-core-api-auth/src/main/java/org/eclipse/che/security/oauth1/OAuthAuthenticator.java:238]: local means that this URL will point to the local authettication REST method, not the remote server auth URL like method above.

Comment (skabashnyuk) [reply wsmaster/che-core-api-auth/src/main/java/org/eclipse/che/security/oauth1/OAuthAuthenticator.java:238]: ok

**Review by skabashnyuk — APPROVED** [review]

**Review by metlos — APPROVED** [review]

**Pull request merged**